cli/src/lib.rs

```diff
@@ -1142,40 +1144,60 @@ pub fn verify_bin(program_id: Pubkey, bin_path: &Path, cluster: &str) -> Result<
             .get_account_with_commitment(&program_id, CommitmentConfig::default())?
             .value
             .map_or(Err(anyhow!("Account not found")), Ok)?;
-        match account.state()? {
-            UpgradeableLoaderState::Program {
-                programdata_address,
-            } => {
-                let account = client
-                    .get_account_with_commitment(&programdata_address, CommitmentConfig::default())?
-                    .value
-                    .map_or(Err(anyhow!("Account not found")), Ok)?;
-                let bin = account.data
-                    [UpgradeableLoaderState::programdata_data_offset().unwrap_or(0)..]
-                    .to_vec();
-
-                if let UpgradeableLoaderState::ProgramData {
-                    slot,
-                    upgrade_authority_address,
-                } = account.state()?
-                {
-                    let state = BinVerificationState::ProgramData {
+        if account.owner == bpf_loader::id() || account.owner == bpf_loader_deprecated::id() {
+            let bin = account.data.to_vec();
+            let state = BinVerificationState::ProgramData {
+                slot: 0, // Need to look through the transaction history.
+                upgrade_authority_address: None,
+            };
+            (bin, state)
+        } else if account.owner == bpf_loader_upgradeable::id() {
+            match account.state()? {
+                UpgradeableLoaderState::Program {
+                    programdata_address,
+                } => {
+                    let account = client
+                        .get_account_with_commitment(
+                            &programdata_address,
+                            CommitmentConfig::default(),
+                        )?
+                        .value
+                        .map_or(Err(anyhow!("Account not found")), Ok)?;
+                    let bin = account.data
+                        [UpgradeableLoaderState::programdata_data_offset().unwrap_or(0)..]
+                        .to_vec();
+
+                    if let UpgradeableLoaderState::ProgramData {
                         slot,
                         upgrade_authority_address,
-                    };
-                    (bin, state)
-                } else {
-                    return Err(anyhow!("Expected program data"));
+                    } = account.state()?
+                    {
+                        let state = BinVerificationState::ProgramData {
+                            slot,
+                            upgrade_authority_address,
+                        };
+                        (bin, state)
+                    } else {
+                        return Err(anyhow!("Expected program data"));
+                    }
+                }
+                UpgradeableLoaderState::Buffer { .. } => {
+                    let offset = UpgradeableLoaderState::buffer_data_offset().unwrap_or(0);
```

Comment:
Surprised that `clippy` did not proposed replace `unwrap_or(0)` to `unwrap_or_default`.

Reply:
Ah interesting, clippy would have complained if it was written `usize::default()` instead of `0` most likely: https://rust-lang.github.io/rust-clippy/master/#unwrap_or_else_default